src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import Any, Optional, Set
+
+import torch
+import torch.nn as nn
+
+from peft.tuners.lycoris_utils import LycorisLayer
+
+
+def parse_positions(positions: str):
+    # Code borrow from https://github.com/stanfordnlp/pyreft/pyreft/dataset.py
+    first_n, last_n = 0, 0
+    if "+" in positions:
+        first_n = int(positions.split("+")[0].strip("f"))
+        last_n = int(positions.split("+")[1].strip("l"))
+    else:
+        if "f" in positions:
+            first_n = int(positions.strip("f"))
+        elif "l" in positions:
+            last_n = int(positions.strip("l"))
+    return first_n, last_n
+
+class LoReftLayer(nn.Module, LycorisLayer):
+    # All names of layers that may contain adapter weights
+    adapter_layer_names = ("reft_A", "reft_R")
+    # other_param_names is defined on parent class
+
+    def __init__(self, base_layer: nn.Module):
+        super().__init__()
+        LycorisLayer.__init__(self, base_layer)
+
+        # ReFT info
+        self.reft_A = nn.ModuleDict({})
+        self.reft_R = nn.ModuleDict({})
+        self.loc = {}
+        self.first_n = {}
+        self.last_n = {}
+        self.dropout = {}
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Conv2d):
+            in_features, out_features = base_layer.in_channels, base_layer.out_channels
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+
+    @property
+    def _available_adapters(self) -> Set[str]:
+        return {*self.reft_A, *self.reft_R}
+
+    def create_adapter_parameters(self, adapter_name: str, r: int):
+        rotate_layer = torch.nn.Linear(self.out_features, r, bias=False)
+        self.reft_R[adapter_name] = torch.nn.utils.parametrizations.orthogonal(rotate_layer, orthogonal_map='cayley')
+        self.reft_A[adapter_name] =  torch.nn.Linear(self.out_features, r)
+
+    def merge(self, safe_merge: bool = False, adapter_names: Optional[list[str]] = None) -> None:
+        pass
```

Comment:
So merging is not supported, right? If we just pass when the user calls `model.merge()`, they might get the impression that it worked. Instead, we should raise a `ValueError` that merging is not possible with LoReFT. Same for `unmerge`.